File: smalivm/src/main/java/org/cf/smalivm/opcode/FillArrayDataPayloadOp.java
package org.cf.smalivm.opcode;

import org.apache.commons.lang3.ClassUtils;
import org.cf.smalivm.context.ExecutionNode;
import org.cf.smalivm.context.HeapItem;
import org.cf.smalivm.context.MethodState;
import org.jf.dexlib2.builder.MethodLocation;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.reflect.Array;
import java.util.List;

public class FillArrayDataPayloadOp extends MethodStateOp {

    @SuppressWarnings("unused")
    private static final Logger log = LoggerFactory.getLogger(FillArrayDataPayloadOp.class.getSimpleName());

    private static Object getProperValue(Number number, Class<?> expectedClass) {
        Class<?> klazz = ClassUtils.wrapperToPrimitive(number.getClass());
        Object value = null;

        // Dexlib will only ever make byte (t), int, long (l), or short (s)
        if (klazz == byte.class) {
            value = number.byteValue();
        } else if (klazz == short.class) {
            value = number.shortValue();
        } else if (klazz == int.class) {
            value = number.intValue();
        } else if (klazz == long.class) {
            value = number.longValue();
        }

        if (expectedClass == boolean.class) {
            value = (byte) value == 1;
        } else if (expectedClass == char.class) {
            value = (char) number.shortValue();
        } else if (expectedClass == short.class) {
            value = number.shortValue();
        } else if (expectedClass == int.class) {
            value = number.intValue();
        } else if (expectedClass == long.class) {
            value = number.longValue();
        } else if (expectedClass == float.class) {
            value = Float.intBitsToFloat(number.intValue());
        } else if (expectedClass == double.class) {
            value = Double.longBitsToDouble(number.longValue());
        }

        return value;
    }

    private final List<Number> arrayElements;
    private final int elementWidth;

    FillArrayDataPayloadOp(MethodLocation location, int elementWidth, List<Number> arrayElements) {
        // childAddress / returnAddress not known until runtime
        super(location);

        this.elementWidth = elementWidth;
        this.arrayElements = arrayElements;
    }

    @Override
    public void execute(ExecutionNode node, MethodState mState) {
        MethodState parent = mState.getParent();
        int targetRegister = parent.getRegistersAssigned()[0];
        // Peek rather than read. This pseudo-instruction shouldn't count as an actual usage for the optimizer.
        HeapItem arrayItem = mState.peekRegister(targetRegister);
        if (!arrayItem.isUnknown()) {
            Object array = arrayItem.getValue();
            Class<?> expectedClass = array.getClass().getComponentType();
            for (int i = 0; i < arrayElements.size(); i++) {
                Number number = arrayElements.get(i);
                Object value = getProperValue(number, expectedClass);
                Array.set(array, i, value);
            }
            // Poke rather than assign for the optimizer.
            mState.pokeRegister(targetRegister, arrayItem);
        }

        MethodLocation returnLocation = mState.getParent().getPseudoInstructionReturnInstruction();
        node.setChildLocations(returnLocation);
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder(".array-data ");
        sb.append(elementWidth);
        for (Number number : arrayElements) {
            sb.append("\n\t").append(number.toString());
        }
        sb.append("\n.end array-data");

        return sb.toString();
    }

}

File: smalivm/src/main/java/org/cf/smalivm/opcode/FillArrayDataPayloadOp.kt
package org.cf.smalivm.opcode

import ExceptionFactory
import org.apache.commons.lang3.ClassUtils
import org.cf.smalivm.configuration.Configuration
import org.cf.smalivm.dex.SmaliClassLoader
import org.cf.smalivm.type.ClassManager
import org.cf.smalivm2.ExecutionNode
import org.cf.smalivm2.OpChild
import org.jf.dexlib2.builder.MethodLocation
import org.jf.dexlib2.iface.instruction.formats.ArrayPayload
import org.slf4j.LoggerFactory
import java.lang.reflect.Array

class FillArrayDataPayloadOp internal constructor(
    location: MethodLocation,
    private val elementWidth: Int,
    private val arrayElements: List<Number>
) :
    Op(location) {

    override val registersReadCount = 2
    override val registersAssignedCount = 1

    override fun execute(node: ExecutionNode): kotlin.Array<out OpChild> {
        val parent = node.state.getParent()!!
        val targetRegister = parent.registersAssigned.toList()[0]
        val array = node.state.peekRegister(targetRegister)!!
        if (array.isKnown) {
            val expectedClass = array.value!!.javaClass.componentType
            for (i in arrayElements.indices) {
                val number = arrayElements[i]
                val value = getProperValue(number, expectedClass)
                Array.set(array.value, i, value)
            }
            node.state.pokeRegister(targetRegister, array)
        }
        // This is a strange pseudo-op which should return to the next op after the FillArrayDataOp.
        val returnLocation = parent.getPsuedoInstructionReturnLocation()
        return collectChildren(returnLocation)
    }

    override fun toString(): String {
        val sb = StringBuilder(".array-data ")
        sb.append(elementWidth)
        for (number in arrayElements) {
            sb.append("\n\t").append(number.toString())
        }
        sb.append("\n.end array-data")
        return sb.toString()
    }

    companion object : OpFactory {
        private val log = LoggerFactory.getLogger(FillArrayDataPayloadOp::class.java.simpleName)

        private fun getProperValue(number: Number, expectedClass: Class<*>): Any {
            val klazz = ClassUtils.wrapperToPrimitive(number.javaClass)

            // Dexlib will only ever make byte (t), int, long (l), or short (s)
            val classValue = when (klazz) {
                Byte::class.javaPrimitiveType -> number.toByte()
                Short::class.javaPrimitiveType -> number.toShort()
                Int::class.javaPrimitiveType -> number.toInt()
                Long::class.javaPrimitiveType -> number.toLong()
                else -> throw IllegalArgumentException("Unexpected array data payload type: $klazz")
            }

            return when (expectedClass) {
                Boolean::class.javaPrimitiveType -> classValue.toInt() == 1
                Char::class.javaPrimitiveType -> number.toShort().toChar()
                Short::class.javaPrimitiveType -> number.toShort()
                Int::class.javaPrimitiveType -> number.toInt()
                Long::class.javaPrimitiveType -> number.toLong()
                Float::class.javaPrimitiveType -> java.lang.Float.intBitsToFloat(number.toInt())
                Double::class.javaPrimitiveType -> java.lang.Double.longBitsToDouble(number.toLong())
                else -> throw IllegalArgumentException("Unexpected array data payload expected type: $expectedClass")
            }
        }

        override fun build(
            location: MethodLocation,
            addressToLocation: Map<Int, MethodLocation>,
            classManager: ClassManager,
            exceptionFactory: ExceptionFactory,
            classLoader: SmaliClassLoader,
            configuration: Configuration
        ): Op {
            val instr = location.instruction as ArrayPayload
            val elementWidth = instr.elementWidth
            val arrayElements = instr.arrayElements
            return FillArrayDataPayloadOp(location, elementWidth, arrayElements)
        }
    }
}
